src/consensus.rs

```diff
@@ -313,14 +312,20 @@ impl <L, M> Consensus<L, M> where L: Log, M: StateMachine {
                         if existing_term != leader_prev_log_term {
                             scoped_debug!("AppendEntriesRequest: inconsistent previous log term: leader: {}, local: {}",
                                           leader_prev_log_term, existing_term);
-                            messages::append_entries_response_inconsistent_prev_entry(self.current_term())
+                            // If an existing entry conflicts with a new one (same index but different terms),
+                            // delete the existing entry and all that follow it
+                            messages::append_entries_response_inconsistent_prev_entry(self.current_term(),
+                                leader_prev_log_index)
```

Comment:
Can you check what happens when a node becomes leader after rebooting and losing in-memory state about it's peers? I **think** it'll send an append request for its inaugural proposal with `leader_prev_log_index=0`, which, if rejected by the follower (for instance because of a log divergence) would let the leader believe that the peer doesn't have a log at all.
If that's true, should special-case `0` so that the leader simply tries the previous log entry in such a case. And either way, would be good to have a test for this.